crates/orderbook/src/orderbook.rs

```diff
@@ -312,9 +319,57 @@ impl Orderbook {
         Ok(())
     }
 
+    async fn get_replaced_order(
+        &self,
+        new_order: &OrderCreation,
+    ) -> Result<Order, ReplaceOrderError> {
+        #[derive(Deserialize)]
+        struct AppData {
+            metadata: Metadata,
+        }
+        #[derive(Deserialize)]
+        #[serde(rename_all = "camelCase")]
+        struct Metadata {
+            replaced_order: OrderUid,
+        }
+
+        let full_app_data = match &new_order.app_data {
+            OrderCreationAppData::Hash { hash } => Cow::Owned(
+                self.app_data
+                    .find(hash)
+                    .await
+                    .map_err(ReplaceOrderError::FindAppData)?
+                    .ok_or(ReplaceOrderError::InvalidAppDataHash)?,
+            ),
+            OrderCreationAppData::Both { full, .. } | OrderCreationAppData::Full { full } => {
+                Cow::Borrowed(full)
+            }
+        };
+
+        let app_data: AppData = serde_json::from_str(full_app_data.as_str())
+            .map_err(ReplaceOrderError::InvalidAppData)?;
+
+        self.find_order_for_cancellation(&app_data.metadata.replaced_order)
+            .await
+            .map_err(Into::into)
+    }
+
+    async fn replace_order_verification(
+        &self,
+        validated_new_order: &Order,
+        validated_old_order: &Order,
+    ) -> Result<(), ReplaceOrderError> {
+        // Verify that the new order is a valid replacement order by checking
+        // that the `app_data` encodes an order cancellation and that both the
+        // old and new orders have the same signer.
+        if validated_new_order.metadata.owner != validated_old_order.metadata.owner {
```

Comment:
nit: I think inlining this part would be fine. Slightly less mental overhead jumping around and also having this part in its own function adds a lot of code (relatively) just for the signature.